server/src/middleware/validate_token.rs

```diff
@@ -19,76 +20,255 @@ pub async fn validate_token(
-                    .map_into_right_body(),
-            };
-            Ok(res)
+
+                TokenValidationStatus::Unknown => Err(EdgeError::AuthorizationDenied),
+                TokenValidationStatus::Invalid => Err(EdgeError::Forbidden("".into())),
+                TokenValidationStatus::Trusted => unreachable!(),
+            }
         }
-        None => {
-            let res = match token_cache.get(&token.token) {
-                Some(t) => {
-                    let token = t.value();
-                    match token.token_type {
-                        Some(TokenType::Client) => {
-                            if req.path().contains("/api/client") {
-                                srv.call(req).await?.map_into_left_body()
-                            } else {
-                                req.into_response(HttpResponse::Forbidden().finish())
-                                    .map_into_right_body()
-                            }
-                        }
-                        Some(TokenType::Frontend) => {
-                            if req.path().contains("/api/frontend")
-                                || req.path().contains("/api/proxy")
-                            {
-                                srv.call(req).await?.map_into_left_body()
-                            } else {
-                                req.into_response(HttpResponse::Forbidden().finish())
-                                    .map_into_right_body()
-                            }
-                        }
-                        None => srv.call(req).await?.map_into_left_body(),
-                        _ => req
-                            .into_response(HttpResponse::Forbidden().finish())
-                            .map_into_right_body(),
-                    }
+        None => match token_cache.get(&token.token) {
+            Some(t) => {
+                let token = t.value();
+                match token.token_type {
+                    Some(TokenType::Frontend) => check_frontend_path(path),
+                    Some(TokenType::Client) => check_backend_path(path),
+                    None => Ok(()),
+                    _ => Err(EdgeError::Forbidden("".into())),
                 }
-                None => req
-                    .into_response(HttpResponse::Forbidden().finish())
-                    .map_into_right_body(),
-            };
+            }
+            None => Err(EdgeError::Forbidden("".into())),
+        },
+    }
+}
+
+fn check_frontend_path(path: &str) -> Result<(), EdgeError> {
+    if path.contains("/api/frontend") || path.contains("/api/proxy") {
+        Ok(())
+    } else {
+        Err(EdgeError::Forbidden("".into()))
+    }
+}
+
+fn check_backend_path(path: &str) -> Result<(), EdgeError> {
+    if path.contains("/api/client") {
+        Ok(())
+    } else {
+        Err(EdgeError::Forbidden("".into()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
New tests, because nothing was covered here

Reply:
I'm impressed you got the compiler to be happy here, I struggled so much with getting it to see the correct body type. Happy that you did though

Reply:
I struggled too, full on wrestling match with the compiler here